tests/test_interpolation.py

```diff
@@ -0,0 +1,164 @@
+    return create_interpolation_data().values
+
+
+class TestInterpolationMethods:
+    ti = 0
+    zi, yi, xi = 1, 1, 1
+
+    @pytest.mark.parametrize(
+        "func, eta, xsi, expected",
+        [
+            pytest.param(interpolation._nearest_2d, 0.49, 0.49, 3.0, id="nearest_2d-1"),
+            pytest.param(interpolation._nearest_2d, 0.49, 0.51, 4.0, id="nearest_2d-2"),
+            pytest.param(interpolation._nearest_2d, 0.51, 0.49, 5.0, id="nearest_2d-3"),
+            pytest.param(interpolation._nearest_2d, 0.51, 0.51, 6.0, id="nearest_2d-4"),
+            pytest.param(interpolation._tracer_2d, None, None, 6.0, id="tracer_2d"),
+            # pytest.param(interpolation._linear_2d, ...),
+            # pytest.param(interpolation._linear_invdist_land_tracer_2d, ...),
+        ],
+    )
+    def test_2d(self, data_2d, func, eta, xsi, expected):
+        ctx = interpolation.InterpolationContext2D(data_2d, eta, xsi, self.ti, self.yi, self.xi)
+        assert func(ctx) == expected
+
+    @pytest.mark.parametrize(
+        "func, eta, xsi, expected",
+        [
+            # pytest.param(interpolation._nearest_3d, ...),
+            # pytest.param(interpolation._cgrid_velocity_3d, ...),
+            # pytest.param(interpolation._linear_invdist_land_tracer_3d, ...),
+            # pytest.param(interpolation._linear_3d, ...),
+            # pytest.param(interpolation._tracer_3d, ...),
+        ],
+    )
+    def test_3d(self, data_3d, func, zeta, eta, xsi, expected):
+        ctx = interpolation.InterpolationContext3D(data_3d, zeta, eta, xsi, self.ti, self.zi, self.yi, self.xi)
+        assert func(ctx) == expected
+
+
+@pytest.mark.usefixtures("tmp_interpolator_registry")
+def test_interpolator_override():
+    fieldset = create_fieldset_zeros_3d()
+
+    @interpolation.register_3d_interpolator("linear")
+    def test_interpolator(ctx: interpolation.InterpolationContext3D):
+        raise NotImplementedError
+
+    with pytest.raises(NotImplementedError):
+        fieldset.U[0, 0.5, 0.5, 0.5]
+
+
+@pytest.mark.usefixtures("tmp_interpolator_registry")
+def test_full_depth_provided_to_interpolators():
+    """The full depth needs to be provided to the interpolation schemes as some interpolators
+    need to know whether they are at the surface or bottom of the water column.
+
+    https://github.com/OceanParcels/Parcels/pull/1816#discussion_r1908840408
+    """
+    xdim, ydim, zdim = 10, 11, 12
+    fieldset = create_fieldset_zeros_3d(xdim=xdim, ydim=ydim, zdim=zdim)
+
+    @interpolation.register_3d_interpolator("linear")
+    def test_interpolator2(ctx: interpolation.InterpolationContext3D):
+        assert ctx.data.shape[1] == zdim
+        # The array z dimension is the same as the fieldset z dimension
+        return 0
+
+    fieldset.U[0.5, 0.5, 0.5, 0.5]
+
+
+@pytest.mark.parametrize("interp_method", ["linear", "freeslip", "nearest", "cgrid_velocity"])
+def test_scipy_vs_jit(interp_method):
```

Comment:
I added this unit test to check if interpolation/integration of Scipy and JIT give the same results (to within a small tolerance). Note that the "cgrid_velocity" interpolation_method currently breaks on my local machine; the other three interpolation methods pass. So it seems that there is a bug somewhere in the "cgrid_velocity"? Perhaps good to first also check this function on current `master`, though

@VeckoTheGecko, feel free to improve/update this unit test!

Reply:
> So it seems that there is a bug somewhere in the "cgrid_velocity"?

I don't think so. From what I can tell `runtime=4e-3, dt=1e-3` are too small timesteps which leads to a loss in precision. The error goes away with `runtime=4, dt=1` which I think would be more reasonable anyway?

Reply:
Well, the test passes with these larger runtime because the `pset` is empty at the end of the execute then. On such a small domain, the velocities are simply too large to keep particles inside the domain long enough. They all get OutofBounds and thus deleted, leaving an empty `pset` (and hence no failing asserts)

Dividing all the velocities by 1000 reintroduces the assert error (see d6f13ad)

Reply:
https://github.com/OceanParcels/Parcels/pull/1834